docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,292 @@
+most restrictive set of constraints.
+1.  Service Account Constraints - security context constraints for service accounts will be created that allows
+pods to run with different permissions than the user.  The user should be required to have
+permissions to reference the service account in order to run a pod with the service account's
+security context constraints.
+1.  Administrator Constraints - a set of constraints with elevated privileges that can be used
+by an administrative user or group.
+1.  Infrastructure Component Constraints - constraints that may have elevated privileges within the
+cluster but possibly not as elevated as administrator constraints.
+
+```go
+// SecurityContextConstraints governs the ability to make requests that affect the SecurityContext that will
+// be applied to a container.
+type SecurityContextConstraints struct {
+  // Type distinguishes the type of SecurityContextConstraints object.  Acceptable values are
+  // INFRA, USER, DEFAULT_USER, SERVICE_ACCOUNT, DEFAULT_SERVICE_ACCOUNT
+  Type string
+  // AllowPrivilegedContainer determines if a container can request to be run as privileged.
+  AllowPrivilegedContainer bool
+  // AllowedCapabilities is a list of capabilities that can be requested to add to the container.
+  AllowedCapabilities []CapabilityType
+  // AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
+  AllowHostDirVolumePlugin bool
+  // The users who have permissions to use this security context constraints
+  AllowedUsers []string
+  // The service accounts that have permission to use this security context constraints
+  AllowedServiceAccounts []NamespacedName
+}
+
+type SecurityContextConstraintsAllocator interface {
+  // Create a SecurityContext based on the given constraints
+  CreateSecurityConstraints(constraints *api.SecurityContextConstraints) api.SecurityContext
+  // Ensure a container's SecurityContext is in compliance with the given constraints
+  ValidateSecurityConstraints(container *api.Container, constraints *api.SecurityContextConstraints) fielderrors.ValidationErrorList
+}
+
+// implements the SecurityContextConstraintsAllocator interface
+type SimpleSecurityContextConstraintsAllocator struct {
+    // SELinuxContext is the strategy that will dictate what labels will be set in the SecurityContext.
+    SELinuxContext SELinuxContextStrategy
+    // RunAsUser is the strategy that will dictate what RunAsUser is used in the SecurityContext.
+    RunAsUser RunAsUserStrategy
+}
+
+const (
+  // container must have SELinux labels of X applied.
+  SELinuxStrategyMustRunAs SELinuxContextStrategyType = "MustRunAs"
+  // container may make requests for any SELinux context labels.
+  SELinuxStrategyRunAsAny SELinuxContextStrategyType = "RunAsAny"
+  // containers must run with the default settings, their requests are ignored
+  SELinuxStrategyRunAsDefault SELinuxContextStrategyType = "RunAsDefault"
```

Comment:
the strategies seems like they belong in the constraints... otherwise, how do you let cluster-admins `RunAsAny` while marking normal users/service-accounts `MustRunAs`?